bench/irmin-pack/tree.ml

```diff
@@ -49,7 +35,148 @@ let pp_store_type ppf = function
+ fun repr channel ->
+  let decode_bin = Repr.decode_bin repr |> Repr.unstage in
+  let decode_prefix = Repr.(decode_bin int32 |> unstage) in
+  let produce_op () =
+    try
+      (* First read the prefix *)
+      let prefix = really_input_string channel 4 in
+      let len', len = decode_prefix prefix 0 in
+      assert (len' = 4);
+      let len = Int32.to_int len in
+      (* Then read the repr *)
+      let content = really_input_string channel len in
+      let len', op = decode_bin content 0 in
+      assert (len' = len);
+      Some (op, ())
+    with End_of_file -> None
+  in
+  Seq.unfold produce_op ()
+
+let seq_mapi64 s =
+  let i = ref Int64.minus_one in
+  Seq.map
+    (fun v ->
+      i := Int64.succ !i;
+      (!i, v))
+    s
+
+(** EMA *)
+module Exponential_moving_average = struct
+  type t = {
+    momentum : float;
+    opp_momentum : float;
+    value : float;
+    count : float;
+  }
+
+  (** [create m] is [ema], a functional exponential moving average. [1. -. m] is
+      the fraction of what's forgotten of the past during each [update].
+
+      When [m = 0.], all the past is forgotten on each [update], i.e. [peek ema]
+      is the latest point fed to [update]. *)
+  let create momentum =
+    if momentum < 0. || momentum >= 1. then invalid_arg "Wrong momentum";
+    { momentum; opp_momentum = 1. -. momentum; value = 0.; count = 0. }
+
+  (** [from_half_life hl] is [ema], a functional exponential moving average.
+      After [hl] calls to [update], half of the past is forgotten. *)
+  let from_half_life hl =
+    if hl < 0. then invalid_arg "Wrong half life";
+    (if hl = 0. then 0. else log 0.5 /. hl |> exp) |> create
+
+  (** [from_half_life_ratio hl_ratio step_count] is [ema], a functional
+      exponential moving average. After [hl_ratio * step_count] calls to
+      [update], half of the past is forgotten. *)
+  let from_half_life_ratio hl_ratio step_count =
+    if hl_ratio < 0. then invalid_arg "Wrong half life ratio";
+    if step_count <= 0L then invalid_arg "Wront step count";
+    Int64.to_float step_count *. hl_ratio |> from_half_life
+
+  (** Feed a new point to the EMA. *)
+  let update ema point =
+    let value = (ema.value *. ema.momentum) +. (point *. ema.opp_momentum) in
+    let count = ema.count +. 1. in
+    { ema with value; count }
+
+  (** Read the EMA value. *)
+  let peek ema =
+    if ema.count = 0. then failwith "Can't peek an EMA before first update";
+    ema.value /. (1. -. (ema.momentum ** ema.count))
+end
+
+module Bootstrap_trace = struct
+  type 'a scope = Forget of 'a | Keep of 'a [@@deriving repr]
+  type key = string list [@@deriving repr]
+  type hash = string [@@deriving repr]
+  type message = string [@@deriving repr]
+
+  type add = {
+    key : key;
+    value : string;
+    ctx0 : int64 scope;
+    ctx1 : int64 scope;
+  }
+  [@@deriving repr]
+
+  type copy = {
+    key_src : key;
+    key_dst : key;
+    ctx0 : int64 scope;
+    ctx1 : int64 scope;
```

Comment:
`ctx0` and `ctx1` are a bit confusing for me - do you mind naming them `hash_id`, `tree_id` or something similar?

Reply:
What about that?
```diff
-    ctx0 : int64 scope;
-    ctx1 : int64 scope;
+    in_ctx_id : int64 scope;
+    out_ctx_id : int64 scope;
```